core/src/main/scala/org/apache/spark/storage/LocalFileSystem.scala

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.storage
+
+import scala.collection.mutable.ArrayBuffer
+import org.apache.hadoop.fs.{LocalFileSystem => HadoopLocalFileSystem}
+import org.apache.hadoop.fs.{Path, PathFilter, RemoteIterator, LocatedFileStatus}
+
+/**
+ * FileSystem for Spark that takes file ordering into account. This is because
+ * Hadoop MapReduce doesn't care about file ordering and HDFS may provide the
+ * file parts in order, but native filesystems may not. As Spark has a notion of
+ * ordering in RDDs (e.g. sortByKey), reading partitions out of order destroys
+ * these notions.
+ *
+ * We only need to override listLocatedStatus as this is called from
+ * FileInputFormat.singleThreadedListStatus and LocatedFileStatusFetcher.
+ */
+class LocalFileSystem extends HadoopLocalFileSystem {
+  override 
+  def listLocatedStatus(path: Path) : RemoteIterator[LocatedFileStatus] = {
+    val listing = super.listLocatedStatus(path)
+    val builder = new ArrayBuffer[LocatedFileStatus]()
+    while(listing.hasNext) {
+      builder += listing.next
+    }
+    val sorted = builder.toArray.sortWith{ (lhs, rhs) => {
```

Comment:
Its easy enough to go make an `Ordering[Path]`, given that its already `Comparable`.  Just because the param *could* be an implicit doesn't mean it *has* to be:

```
scala> val a = Array[Path]()
a.sorted(new Ordering[Path]{def compare(x:Path, y:Path): Int = x.compareTo(y)})
res4: Array[org.apache.hadoop.fs.Path] = Array()
```

There is an automatic conversion from `Comparable[T]` to `Ordering[T]` -- I think the problem is that `Path` is un-parameterized `Comparable`:

```
scala> abstract class Q extends Comparable[Q]
defined class Q
scala> implicitly[Ordering[Q]]
res0: Ordering[Q] = scala.math.LowPriorityOrderingImplicits$$anon$6@48a22925
```

anyhow, probably doesn't effect the code much in this case